sdk/core/azure-core/src/main/java/com/azure/core/util/Context.java

```diff
@@ -256,15 +265,21 @@ public Optional<Object> getData(Object key) {
      * @return A map containing all values of the context linked-list.
      */
     public Map<Object, Object> getValues() {
-        return getValuesHelper(new HashMap<>());
-    }
+        Map<Object, Object> map = new HashMap<>((int) Math.ceil(contextCount / 0.75F));
 
-    private Map<Object, Object> getValuesHelper(Map<Object, Object> values) {
-        if (key != null) {
-            values.putIfAbsent(key, value);
+        for (Context pointer = this; pointer != null; pointer = pointer.parent) {
+            if (pointer.key != null) {
+                map.putIfAbsent(pointer.key, pointer.value);
+            }
+
+            // If the contextCount is 1 that means the next parent Context is the NONE Context.
+            // Break out of the loop to prevent a meaningless check.
+            if (pointer.contextCount == 1) {
+                break;
+            }
         }
 
-        return (parent == null) ? values : parent.getValuesHelper(values);
+        return map;
```

Comment:
For a given Context instance, this `map` will always be the same as `Context` is immutable. So, we can reuse this map when `getValues()` is called a second time. We just have to return `Collections.unmodifiableMap(map)`

Reply:
I'll make this into a deferred local variable, given the APIs low usage I won't make the property volatile to reduce memory synchronization